src/test/spec/retryable_reads.rs

```diff
@@ -0,0 +1,183 @@
+                let can_run_on = run_on.iter().any(|run_on| {
+                    if let Some(ref min_server_version) = run_on.min_server_version {
+                        let (major, minor) = parse_version(min_server_version);
+                        if client.server_version_lt(major, minor) {
+                            return false;
+                        }
+                    }
+                    if let Some(ref max_server_version) = run_on.max_server_version {
+                        let (major, minor) = parse_version(max_server_version);
+                        if client.server_version_gt(major, minor) {
+                            return false;
+                        }
+                    }
+                    if let Some(ref topology) = run_on.topology {
+                        if !topology.contains(&client.topology()) {
+                            return false;
+                        }
+                    }
+                    true
+                });
+                if !can_run_on {
+                    println!("Skipping {}", test_case.description);
+                    continue;
+                }
+            }
+
+            match test_case.fail_point {
+                Some(_) => {
+                    let _guard = LOCK.run_exclusively().await;
+                }
+                _ => {
+                    let _guard = LOCK.run_concurrently().await;
+                }
+            }
+
+            let db_name = match test_file.database_name {
+                Some(ref db_name) => db_name.clone(),
+                None => test_case.description.replace('$', "%").replace(' ', "_"),
+            };
+
+            let coll_name = match test_file.collection_name {
+                Some(ref coll_name) => coll_name.clone(),
+                None => String::from("coll"),
+            };
+
+            let coll = client.init_db_and_coll(&db_name, &coll_name).await;
+
+            if let Some(ref data) = test_file.data {
+                let data: Vec<Document> = data
+                    .as_array()
+                    .unwrap()
+                    .iter()
+                    .map(|doc| doc.as_document().unwrap().clone())
+                    .collect();
+                if !data.is_empty() {
+                    coll.insert_many(data, None)
+                        .await
+                        .expect(&test_case.description);
+                }
+            }
+
+            if let Some(ref fail_point) = test_case.fail_point {
+                client
+                    .database("admin")
+                    .run_command(fail_point.clone(), None)
+                    .await
+                    .unwrap();
+            }
+
+            let mut events: Vec<TestEvent> = Vec::new();
+            for operation in test_case.operations {
+                let test_operation: AnyTestOperation =
+                    bson::from_bson(Bson::Document(operation.as_document())).unwrap();
```

Comment:
I tried this out, but it's a bit tricky since `TestCase` derives `Debug`. `AnyTestOperation` has a field `operation: Box<dyn TestOperation>`, so as far as I know it can't derive `Debug` since `TestOperation` is a trait which means not all of `AnyTestOperation`'s fields derive `Debug`. Is there a good workaround for this?

Reply:
Could you update the `operation` field to be `operation: Box<dyn TestOperation + Debug>`? Alternately, since this is only an internal API, we could define `TestOperation` so that it requires `Debug`, i.e. `trait TestOperation: Debug`.